xbmc/windowing/wayland/WinSystemWayland.cpp

```diff
@@ -466,89 +553,207 @@ bool CWinSystemWayland::SetFullScreen(bool fullScreen, RESOLUTION_INFO& res, boo
+  {
+    // Buffer scale must also match egl size configuration
+    ApplyBufferScale();
+  }
+  if (update.surfaceSizeChanged)
   {
-    g_application.ReloadSkin();
+    // Mark everything opaque so the compositor can render it faster
+    // Do it here so size always matches the configured egl surface
+    CLog::LogF(LOGDEBUG, "Setting opaque region size %dx%d", m_surfaceSize.Width(), m_surfaceSize.Height());
+    wayland::region_t opaqueRegion = m_compositor.create_region();
+    opaqueRegion.add(0, 0, m_surfaceSize.Width(), m_surfaceSize.Height());
+    m_surface.set_opaque_region(opaqueRegion);
   }
+  if (update.configuredSizeChanged)
+  {
+    // Update window decoration state
+    m_windowDecorator->SetState(m_configuredSize, m_scale, m_shellSurfaceState);
+    m_shellSurface->SetWindowGeometry(m_windowDecorator->GetWindowGeometry());
+  }
+  // Set always, because of initialization order GL context has to keep track of
+  // whether the size changed. If we skip based on update.bufferSizeChanged here,
+  // GL context will never get its initial size set.
+  SetContextSize(m_bufferSize);
 }
-void CWinSystemWayland::HandleSurfaceConfigure(std::uint32_t serial, std::int32_t width, std::int32_t height)
+
+void CWinSystemWayland::ProcessMessages()
 {
-  CSingleLock lock(g_graphicsContext);
-  CLog::LogF(LOGDEBUG, "Configure serial %u: size %dx%d", serial, width, height);
-  m_currentConfigureSerial = serial;
-  if (!ResetSurfaceSize(width, height, m_scale))
+  Actor::Message* message{};
+  MessageHandle lastConfigureMessage;
+  int skippedConfigures = -1;
+  int newScale = m_scale;
+
+  while (m_protocol.ReceiveOutMessage(&message))
   {
-    // nothing changed, ack immediately
-    AckConfigure(serial);
+    MessageHandle guard{message};
+    switch (message->signal)
+    {
+      case WinSystemWaylandProtocol::CONFIGURE:
+        // Do not directly process configures, get the last one queued:
+        // While resizing, the compositor will usually send a configure event
+        // each time the mouse moves without any throttling (i.e. multiple times
+        // per rendered frame).
+        // Going through all those and applying them would waste a lot of time when
+        // we already know that the size is not final and will change again anyway.
+        skippedConfigures++;
+        lastConfigureMessage = std::move(guard);
+        break;
+      case WinSystemWaylandProtocol::OUTPUT_HOTPLUG:
+        // Maybe the output that was added was the one we should be on?
+        if (m_bFullScreen)
+        {
+          CLog::LogF(LOGDEBUG, "Output hotplug in full screen mode: Re-reading resolutions");
+          UpdateResolutions();
+          // This will call SetFullScreen(), which will match the output against
+          // the information from the resolution and call set_fullscreen on the
+          // surface if it changed.
+          g_graphicsContext.SetVideoResolution(g_graphicsContext.GetVideoResolution(), true);
```

Comment:
We can try that if you are OK with the extra parameter on SetVideoResolution staying there. It will still be necessary, the difference is that windowing will not call it.